command/agentproxyshared/cache/lease_cache.go

```diff
@@ -237,20 +272,54 @@ func (c *LeaseCache) Send(ctx context.Context, req *SendRequest) (*SendResponse,
 		// Cleanup on the cache if there are no remaining inflight requests.
 		// This is the last step, so we defer the call first
 		if inflight != nil && inflight.remaining.Load() == 0 {
-			c.inflightCache.Delete(id)
+			c.inflightCache.Delete(dynamicSecretCacheId)
+			c.inflightCache.Delete(staticSecretCacheId)
 		}
 	}()
 
-	idLock := locksutil.LockForKey(c.idLocks, id)
+	idLockDynamicSecret := locksutil.LockForKey(c.idLocks, dynamicSecretCacheId)
 
 	// Briefly grab an ID-based lock in here to emulate a load-or-store behavior
 	// and prevent concurrent cacheable requests from being proxied twice if
 	// they both miss the cache due to it being clean when peeking the cache
 	// entry.
-	idLock.Lock()
-	inflightRaw, found := c.inflightCache.Get(id)
+	idLockDynamicSecret.Lock()
+	inflightRaw, found := c.inflightCache.Get(dynamicSecretCacheId)
+	if found {
+		idLockDynamicSecret.Unlock()
+		inflight = inflightRaw.(*inflightRequest)
+		inflight.remaining.Inc()
+		defer inflight.remaining.Dec()
+
+		// If found it means that there's an inflight request being processed.
+		// We wait until that's finished before proceeding further.
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case <-inflight.ch:
+		}
+	} else {
+		inflight = newInflightRequest()
+		inflight.remaining.Inc()
+		defer inflight.remaining.Dec()
+
+		c.inflightCache.Set(dynamicSecretCacheId, inflight, gocache.NoExpiration)
```

Comment:
I'm not very familiar with this func, but isn't this else branch something we should only do if we're working on a dynamic secret?

Reply:
We need this for static secrets too for concurrent access -- we add the secret ID to the inflight cache so that other requests can read it from the cache

Reply:
Ok, so here's what I think this code is about: we can't know whether we're dealing with a static or dynamic secret yet, because that's only knowable once we've seen the response, and we don't yet know if we're even going to be talking to Vault.  So we populate two inflight cache items, even though only one is going to be pertinent.  If this is correct, could you add a comment to this effect to the code?

Also, I think there's a bug: we overwrite `inflight`, so the defer that's meant to clean it up when inflight.remaining.Load() is zero will only clean up one of them, in the case where we created two.

Reply:
That's correct! I'll add a comment, and look into the mentioned bug